test/Dialect/FIRRTL/layers-basic.mlir

```diff
@@ -0,0 +1,180 @@
+// RUN: circt-opt %s
+
+firrtl.circuit "layers-basic" {
+
+  //===--------------------------------------------------------------------===//
+  // Props
+  //===--------------------------------------------------------------------===//
+
+  firrtl.layer @A bind {
+    firrtl.layer @B bind {
+    }
+  }
+
+  firrtl.layer @B bind {}
+
+  firrtl.module @Foo(out %o : !firrtl.probe<uint<1>, @A>) {
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @A>
+    firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @A>    
+  }
+
+  firrtl.module @Bar(out %o : !firrtl.probe<uint<1>, @B>) {
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @B>
+    firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @B>
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Basic Casting Tests
+  //===--------------------------------------------------------------------===//
+
+  firrtl.module @CastToAnyLayer() {
+    // Safe to take an uncolored probe and colorize it.
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @A>
+  }
+
+  firrtl.module @CastToSublayer() {
+    // Safe to take a colored probe and cast it to a nested/child/sub-layer.
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @A>
+    %t = firrtl.ref.cast %s : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>, @A::@B>
+  }
+
+  firrtl.module @CastToAmbientLayer(out %o : !firrtl.probe<uint<1>, @A::@B>) {
+    // safe to take a value under layerblock @A and cast it to @A::B
+    firrtl.layerblock @A {  
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @A::@B>
+      firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @A::@B>
+    }
+  }
+
+  firrtl.module @UncoloredDefineUnderLayerblock(out %o : !firrtl.probe<uint<1>>) {
+    // Safe to define an uncolored ref under a layerblock, as long as the define
+    // would be valid outside the layerblock..
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
```

Comment:
I don't think this should be allowed given how defines work (there can be only one define to a probe in the entire module). The stricter check here should be that the ambient layer of a define must exactly match the explicit layer of the destination probe type.

I realize that this conflicts with the code that already landed. However, I've tried to really restrict this in the FIRRTL spec PR.

Reply:
We now check that the ref.define op doesn't have _more_ layer requirements than the destination.